open-telemetry/opentelemetry-android
# Http/sURLConnection auto instrumentation

### What's covered in this PR
As the Android SDK classes can't be instrumented with ByteBuddy Android Gradle Plugin, this PR instruments all calls to relevant URLConnection/HttpURLConnection/HttpsURLConnection APIs in android applications source code. 

**APIs Instrumented:**

**APIs that connect _but do not_ read bytes:**
- connect()
- getOutputStream()

**APIs that connect and read bytes:**
- getContent()
- getContent(Class[])
- getContentEncoding()
- getContentLength()
- getContentType()
- getDate()
- getExpiration()
- getLastModified()
- getHeaderFieldDate(name, Default)
- getHeaderFields()
- getHeaderFieldKey(int)
- getHeaderField(int)
- getHeaderField(name)
- getHeaderFieldInt(name, Default)
- getResponseMessage()
- getResponseCode()
- getInputStream()
- getErrorStream()
- getContentLengthLong() ( added in API level 24)
- getHeaderFieldLong() ( added in API level 24)

**Approach Taken**
- Request header can only be added before connecting. So, trace context header should be added before any of the above API calls are made.
- All above API calls are replaced with our own replacement methods which do the following:
1. Add connection to our ongoing URLConnections map if not added already. If it is the first time connecting, trace context is added as a header to the outgoing request
2. Original API call is made. If method can throw an exception, we wrap the method around a try/catch, so we can close the ongoing spans with exception reporting before re-throwing exception back to client. 
3. We update the last seen timestamp for the connection. If API is amongst the ones that read bytes also, we mark this URLConnection as harvestable.

Harvestable meaning here - if harvestable and if not already reported and if left idle for more than CONNECTION_KEEP_ALIVE_INTERVAL ms, connection is reported by a connection killer thread. So, once connection is made and any data is read from the connection, we mark it as harvestable because we don't anticipate connection to be sitting idle for longer after this.

As noted above, We keep track of all ongoing URLConnections and three other properties for it - last seen time, harvestable (boolean), reported (boolean) 

**When are spans closed and reported:** 
- If any exception is thrown anywhere, spans are closed and reported. 
- As a general behavior, either getInputStream or getErrorStream is always called and called in the end. We close spans and report when these two methods are called. 
- A connection reporter is scheduled which periodically check if any tracked connection, which is marked harvestable and not reported yet, was last seen more than CONNECTION_KEEP_ALIVE_INTERVAL ms before, then it closes the spans and reports the connection. [Currently, couldn't think of a way to proactively schedule this for clients, customers can manually call our API to schedule this runnable for now.]

**Sample App to test this PR**
Added an android sample app with many test scenarios for this PR [here](https://github.com/surbhiia/HttpURLConnection). 

If you would like to test this PR using the above sample app:
- Pull this PR in your local repo.
- Update the return value to some dummy value for the API you would like to test in HttpUrlReplacements.java
- Update gradle.properties with relevant version number
- Use ./gradlew publishToMavenLocal

In sample app:
- Update the version of httpURLConnection agent and lib dependencies to the version number you just published
- Build and run the app. Click the "ClickMe" button. 
- See output in logcat.
